databuilder/codes.py

```diff
@@ -140,35 +140,53 @@ def __post_init__(self):
                 setattr(self, name, category_map)
 
 
-def codelist_from_csv(filename, column, system):
+def codelist_from_csv(filename, column, system=None, category_column=None):
     filename = Path(filename)
     if not filename.exists():
         raise CodelistError(f"No CSV file at {filename}")
     with filename.open("r") as f:
-        return codelist_from_csv_lines(f, column, system)
+        return codelist_from_csv_lines(
+            f, column, system=system, category_column=category_column
+        )
 
 
-def codelist_from_csv_lines(lines, column, system):
-    try:
-        code_class = REGISTRY[system]
-    except KeyError:
-        raise CodelistError(
-            f"No system matching '{system}', allowed are: "
-            f"{', '.join(REGISTRY.keys())}"
-        )
+def codelist_from_csv_lines(lines, column, system=None, category_column=None):
+    if system is None:
+        code_class = str
+    else:
+        if category_column is not None:
+            raise CodelistError(
+                "`system` and `category_column` cannot be supplied together"
+            )
+        try:
+            code_class = REGISTRY[system]
+        except KeyError:
+            raise CodelistError(
+                f"No system matching '{system}', allowed are: "
+                f"{', '.join(REGISTRY.keys())}"
+            )
     # `restval` ensures we never get None instead of string, so `.strip()` will
     # never blow up
     reader = csv.DictReader(iter(lines), restval="")
     if column not in reader.fieldnames:
         raise CodelistError(f"No column '{column}' in CSV")
-    codes = set()
+    codes = []
     # We treat every other column in the CSV as being a mapping from codes to categories
     category_maps = {name: {} for name in reader.fieldnames if name != column}
+    if category_column is not None and category_column not in category_maps:
+        raise CodelistError(f"No column '{category_column}' in CSV")
     for row in reader:
         code_str = row[column].strip()
         if code_str:
             code = code_class(code_str)
-            codes.add(code)
+            codes.append(code)
```

Comment:
If there are duplicate codes in the csv that's passed in, we'll keep them here (and currently get rid of them at the end, when we pass a set into `Codelist`).  If the plan is that we'll get rid of the `Codelist` class and just return a list of codes, would it be better to use a set here, and (once the `Codelist` class is gone) convert to a list when returning?  Otherwise we'll make a list, (probably) turn it into a set to get rid of dupliclates, and then back into a list to return it to the user.

Reply:
I wanted to avoid a set here because I thought it would be a bit more user-friendly to preserve the order. And I've got a separate PR (which I can push as a draft but we can't merge until all the studies have been updated) which refactors all this stuff away in any case. I agree it's not the nicest way of doing things, but I didn't think it was worth sweating over given the code will be short lived.